pkg/core/state/notification_event.go

```diff
@@ -20,10 +20,9 @@ type NotificationEvent struct {
 	Item       *stackitem.Array `json:"state"`
 }
 
-// AppExecResult represent the result of the script execution, gathering together
+// Execution represents the result of a single script execution,  gathering together
 // all resulting notifications, state, stack and other metadata.
-type AppExecResult struct {
-	TxHash         util.Uint256
```

Comment:
I think we have to separate internal `AppExecResult` from RPC result format. Yeah, it's not that nice, but I dont't see how we can solve this problem otherwise. `TxHash` can be renamed to `Containter` then and `AppExecResult` logic stay as is for the most part (except storing two AERs for block).

Reply:
I tried to leave `AppExecResult` untouched, but then I realised that at the current master we're storing application results for block with the help of block hash, and if we're about to keep it the same way, we need to include multiple execution results into `AppExecResult`. So this format fits fine for `getapplicationlog` RPC response, and the only thing that remains is application logs for notifications subsystem. So I came to the desision to use `AppExecResult` with `Executions` field for storing and for `getapplicationlog` RPC, and use new `ApplicationLog` structure for RPC notifications subsistem. `ApplicationLog` represents the single execution result, which allow us to avoid event sequencing problem.

The other possible way is not to modify `AppExecResult` at all, but then we have to modify the way of storing `AppExecResult` in order to allow multiple execution results for single hash. I'm not sure that I know the correct way to implement it.

Reply:
So could you take a look at the current implementation and decide whether we can include multiple executions into `AppExecResult` or it would be better to add some new `dao` method for storing and retrieving application logs for `onPersist` and `postPersist` execution results?

Reply:
How about this?
```
@@ -29,10 +30,11 @@ var (
 // DAO is a data access object.
 type DAO interface {
        AppendNEP5Transfer(acc util.Uint160, index uint32, tr *state.NEP5Transfer) (bool, error)
+       AppendAppExecResult(aer *state.AppExecResult, buf *io.BufBinWriter) error
        DeleteContractState(hash util.Uint160) error
        DeleteStorageItem(id int32, key []byte) error
        GetAndDecode(entity io.Serializable, key []byte) error
-       GetAppExecResult(hash util.Uint256) (*state.AppExecResult, error)
+       GetAppExecResults(hash util.Uint256, trig trigger.Type) ([]state.AppExecResult, error)
        GetBatch() *storage.MemBatch
        GetBlock(hash util.Uint256) (*block.Block, error)
        GetContractState(hash util.Uint160) (*state.Contract, error)
```

Reply:
Now there are just a few changes into `blockchain` and `dao`, but there's a lot of code duplication between `response.ApplicationLog` and `state.AppExecResult` JSON marshalling. However, we cannot reuse marshallers from `response` due to the import cycle problem.

Reply:
We can probably move some common code out if we need JSON for AER at all.

Reply:
We need JSON for AER for the notifications subsystem. I moved `Execution` struct out of the `result` package, it solves the problem.